app/javascript/packages/session/requests.ts

```diff
@@ -1,44 +1,96 @@
-import { request } from '@18f/identity-request';
+import { request, ResponseError } from '@18f/identity-request';
 
-export interface SessionStatusResponse {
+export interface SessionLiveStatusResponse {
   /**
    * Whether the session is still active.
    */
-  live: boolean;
+  live: true;
 
   /**
    * ISO8601-formatted date string for session timeout.
    */
   timeout: string;
 }
 
-export interface SessionStatus {
+export interface SessionTimedOutStatusResponse {
   /**
    * Whether the session is still active.
    */
-  isLive: boolean;
+  live: false;
 
   /**
    * ISO8601-formatted date string for session timeout.
    */
-  timeout: string;
+  timeout: null;
+}
+
+type SessionStatusResponse = SessionLiveStatusResponse | SessionTimedOutStatusResponse;
+
+interface SessionLiveStatus {
+  /**
+   * Whether the session is still active.
+   */
+  isLive: true;
+
+  /**
+   * ISO8601-formatted date string for session timeout.
+   */
+  timeout: Date;
+}
+
+interface SessionTimedOutStatus {
+  /**
+   * Whether the session is still active.
+   */
+  isLive: false;
+
+  /**
+   * ISO8601-formatted date string for session timeout.
+   */
+  timeout?: undefined;
 }
 
+export type SessionStatus = SessionLiveStatus | SessionTimedOutStatus;
```

Comment:
how come have both `SessionStatus` and `SessionStatusResponse`? I feel like having the union type is useful for the actual TS code, since the code is already mapping responses from the server explicitly, having the union type for `SessionStatusResponse` seems like it doesn't give us much?

Reply:
It was probably useful when I was trying different implementations to get the type narrowing to cooperate with different generic forms. I suspect I left it mostly as a nice symmetry between the "input" and "output" types. Ultimately though, yeah, it's unused, and if anything [the overloaded function generic using it](https://github.com/18F/identity-idp/blob/d1e9f8351eae2e8660bb96f847dca23b9d0f024a/app/javascript/packages/session/requests.ts#L64) would be clearer by just spelling out the union of overloaded forms explicitly.

Reply:
yeah if I was writing this I'd probably collapse the "raw" server types into a singular one, and use the union type for the one passed around in TS code

Reply:
Hm, I think I may have initially misunderstood the suggestion. But after digging in, both interpretations lead me to think that maybe it's worth keeping as-is:

- If the idea is to remove `SessionStatusResponse` as a union of `SessionLiveStatusResponse | SessionTimedOutStatusResponse`, while keeping the latter types, this makes sense for the overloaded method signature, but makes the [`request` generic parameter below](https://github.com/18F/identity-idp/blob/d1e9f8351eae2e8660bb96f847dca23b9d0f024a/app/javascript/packages/session/requests.ts#L91) a little less expressive.
- If the idea is to flatten `SessionStatusResponse` as the interface and update property types to reflect all forms (`live: boolean; timeout: string | null`), this makes TypeScript unable to do type narrowing in the logic of the mapping function, specifically on calling the `Date` constructor with a possibly-null `timeout` ([screenshot](https://github.com/18F/identity-idp/assets/1779930/e76bcd0b-fd3e-43e4-ba8a-2ae56eb854e4)).

As a compromise, I'd suggest keeping the union type for the `request` calls, and updating the overloaded signature to re-reference the individual response types, to reflect the overloaded forms. See 0b0efb519.

Reply:
yeah I had intended the "flatten SessionStatusResponse as the interface and update property types to reflect all forms" approach, but the screenshot makes a good case for keeping. Thanks for trying!